subprojects/core-api/src/main/java/org/gradle/api/artifacts/DependencySubstitutions.java

```diff
@@ -102,9 +103,41 @@
          */
         Substitution because(String reason);
 
+        /**
+         * Specifies that the substituted target dependency should use the specified classifier.
+         *
+         * This method assumes that the target dependency is a jar (type jar, extension jar).
+         *
+         * @since 6.6
+         */
+        @Incubating
+        Substitution usingClassifier(String classifier);
+
+        /**
+         * Specifies that the substituted dependency mustn't have any classifier.
+         * It can be used whenever you need to substitute a dependency which uses a classifier into
+         * a dependency which doesn't.
+         *
+         * This method assumes that the target dependency is a jar (type jar, extension jar).
+         *
+         * @since 6.6
+         */
+        @Incubating
+        Substitution withoutClassifier();
+
```

Comment:
:o: I think we should we also add `withType()` and `withoutType()` while we are at it. You can use `<type>` in Maven to say that you want an artifact with a different file extension than "jar". This is for example used in the Android ecosystem to depend on "aar"s. 
If I am not mistaken, this will lead to both _type_ and _extension_ set to that value in the underlying IvyArtifactName.

Reply:
We have issues opened around this topic. I don't think we should do it mixed with this set of changes. However it might be an interesting thing to consider indeed.

We have to remember that in a dependency declaration, you cannot really specify the Maven `type`, instead you can specify an `ext`. However, the moment you do that, for Gradle you are referencing an `artifact` and no longer a dependency. Gradle effectively ignores all the metadata.

Reply:
Note that you can substitute with type/ext using the lower level API. The `classifier` use case being the most frequent, it's the only one I added as a higher level API, considering we don't want to encourage its use. If you feel strongly I should add `withType`, then I can do it but I would also, in this case, add `withExtension`.

Reply:
It's fine for me if we keep it like it is for now.